File: examples/my-first-galgame/generate_meta.mjs
import fs from 'fs';
import path from 'path';

// 模拟 gg-meta 的功能，生成 meta 文件
function generateMetaFile(filePath) {
    const stats = fs.statSync(filePath);
    const size = stats.size;
    const name = path.basename(filePath);
    const relativePath = path.relative('e:\\灵之镜有限公司\\gg-game-engine', filePath);
    const assetType = getAssetType(name);
    
    // 生成 UUID v7 格式的 GUID
    const guid = generateUUIDv7();
    const timestamp = new Date().toISOString();
    
    // 构建 meta 内容
    const metaContent = `MetaFile({
    version: "1.0",
    asset: Asset({
        type: "${assetType}",
        path: "${relativePath.replace(/\\/g, '/')}",
        guid: "${guid}",
        name: "${name}",
        size: ${size},
        modified: "${timestamp}",
    }),
    import_settings: None,
    dependencies: [],
    references: [],
    timestamp: "${timestamp}",
    hash: None,
})`;
    
    const metaPath = filePath + '.meta';
    fs.writeFileSync(metaPath, metaContent);
    console.log(`Generated meta file for ${filePath}`);
}

function getAssetType(filename) {
    const ext = path.extname(filename).toLowerCase();
    switch (ext) {
        case '.toml':
            return 'config';
        case '.gscript':
            return 'script';
        default:
            return 'unknown';
    }
}

// 生成 UUID v7
function generateUUIDv7() {
    const now = Date.now();
    const random = Math.floor(Math.random() * 0x10000000000000000).toString(16).padStart(16, '0');
    const timestamp = Math.floor(now / 1000).toString(16).padStart(12, '0');
    return `${timestamp.substring(0, 8)}-${timestamp.substring(8, 12)}-7${random.substring(1, 4)}-${(parseInt(random.substring(4, 5), 16) | 0x8).toString(16)}${random.substring(5, 8)}-${random.substring(8).padEnd(12, '0')}`;
}

// 处理目录中的文件
const directory = 'e:\\灵之镜有限公司\\gg-game-engine\\examples\\my-first-galgame';

// 处理 game.toml
const gameTomlPath = path.join(directory, 'game.toml');
if (fs.existsSync(gameTomlPath)) {
    generateMetaFile(gameTomlPath);
}

// 处理 scripts 目录
const scriptsDir = path.join(directory, 'scripts');
if (fs.existsSync(scriptsDir)) {
    const scriptFiles = fs.readdirSync(scriptsDir);
    for (const file of scriptFiles) {
        // 跳过 .meta 文件
        if (file.endsWith('.meta')) {
            continue;
        }
        const scriptPath = path.join(scriptsDir, file);
        if (fs.statSync(scriptPath).isFile()) {
            generateMetaFile(scriptPath);
        }
    }
}

console.log('Meta files generated successfully!');

File: projects/core/gg-meta/examples/generate_meta.rs
use gg_meta::{MetaFile, generate_guid};
use std::path::Path;
use std::fs;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // 目标目录
    let target_dir = Path::new("../../examples/my-first-galgame");
    
    // 处理 game.toml
    let game_toml_path = target_dir.join("game.toml");
    if game_toml_path.exists() {
        generate_meta_for_file(&game_toml_path, "config")?;
    }
    
    // 处理 scripts 目录
    let scripts_dir = target_dir.join("scripts");
    if scripts_dir.exists() && scripts_dir.is_dir() {
        for entry in fs::read_dir(scripts_dir)? {
            let entry = entry?;
            let entry_path = entry.path();
            if entry_path.is_file() && !entry_path.extension().map_or(false, |ext| ext == "meta") {
                generate_meta_for_file(&entry_path, "script")?;
            }
        }
    }
    
    Ok(())
}

fn generate_meta_for_file(file_path: &Path, asset_type: &str) -> Result<(), Box<dyn std::error::Error>> {
    let stats = fs::metadata(file_path)?;
    let size = stats.len();
    let name = file_path.file_name().unwrap().to_string_lossy().to_string();
    let relative_path = get_relative_path(file_path)?;
    
    // 使用 gg-meta 库创建 MetaFile
    let mut meta = MetaFile::new(
        asset_type,
        &relative_path,
        &name,
        size,
    );
    
    // 更新时间戳
    meta.update_timestamp();
    
    // 生成 meta 文件路径
    let meta_path = file_path.with_extension(format!("{}.meta", file_path.extension().unwrap_or_default().to_string_lossy()));
    
    // 写入文件
    meta.to_file(&meta_path)?;
    println!("Generated meta file for {}", file_path.display());
    
    Ok(())
}

/// 获取相对路径
fn get_relative_path(file_path: &Path) -> Result<String, Box<dyn std::error::Error>> {
    let project_root = Path::new("../../../");
    let absolute_path = std::env::current_dir()?.join(file_path);
    let relative_path = absolute_path.strip_prefix(project_root)?;
    Ok(relative_path.to_string_lossy().replace('\\', "/"))
}
